compiler/packages/babel-plugin-react-compiler/src/Inference/DropManualMemoization.ts

```diff
@@ -390,6 +397,36 @@ export function dropManualMemoization(func: HIRFunction): void {
             manualMemo.kind,
             sidemap,
           );
+
+          /**
+           * Bailout on void return useMemos. This is an anti-pattern where code might be using
+           * useMemo like useEffect: running arbirtary side-effects synced to changes in specific
+           * values.
+           */
+          if (
+            func.env.config.enableValidateNoVoidUseMemo &&
+            manualMemo.kind === 'useMemo'
+          ) {
+            const funcToCheck = sidemap.functions.get(
+              fnPlace.identifier.id,
+            )?.value;
+            if (funcToCheck !== undefined && funcToCheck.loweredFunc.func) {
+              if (doesFunctionHaveVoidReturn(funcToCheck.loweredFunc.func)) {
+                errors.push({
+                  severity: ErrorSeverity.InvalidReact,
+                  reason: `React Compiler has skipped optimizing this component because ${
+                    manualMemo.loadInstr.value.kind === 'PropertyLoad'
+                      ? 'React.useMemo'
+                      : 'useMemo'
+                  } doesn't return a value. useMemo should only be used for memoizing values, not running arbitrary side effects.`,
+                  loc: instr.value.loc,
+                  suggestions: null,
+                  description: null,
+                });
```

Comment:
nit: let's use the new `errors.pushDiagnostic(CompilerDiagnostic.create(...))` API instead? You'll get nicer output.